File: src/main/java/com/github/zyypj/booter/minecraft/tool/CooldownController.java
package com.github.zyypj.booter.minecraft.tool;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import org.bukkit.entity.Player;

/**
 * Controlador de Cooldowns para jogadores no Bukkit.
 * Permite criar, verificar e gerenciar cooldowns associados a jogadores.
 */
public class CooldownController {

    private final HashMap<String, Long> cooldowns = new HashMap<>();
    private static final Map<UUID, CooldownController> CONTROLLER = new HashMap<>();

    /**
     * Cria um cooldown associado a uma chave.
     *
     * @param key     A chave do cooldown.
     * @param seconds A duração do cooldown em segundos.
     */
    public void createCooldown(String key, int seconds) {
        this.cooldowns.put(key, System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(seconds));
    }

    /**
     * Remove um cooldown associado a uma chave.
     *
     * @param key A chave do cooldown a ser removido.
     */
    public void deleteCooldown(String key) {
        this.cooldowns.remove(key);
    }

    /**
     * Obtém o tempo restante de um cooldown associado a uma chave.
     *
     * @param key A chave do cooldown.
     * @return O tempo restante em milissegundos ou 0 se o cooldown não existir.
     */
    public long getCooldown(String key) {
        return this.cooldowns.getOrDefault(key, 0L) - System.currentTimeMillis();
    }

    /**
     * Verifica se um cooldown associado a uma chave ainda está ativo.
     *
     * @param key A chave do cooldown.
     * @return true se o cooldown estiver ativo, caso contrário false.
     */
    public boolean isInCooldown(String key) {
        return this.getCooldown(key) > 0;
    }

    /**
     * Remove o controlador de cooldowns associado a um jogador.
     *
     * @param player O jogador cujo controlador será removido.
     */
    public static void removeCooldownController(Player player) {
        CONTROLLER.remove(player.getUniqueId());
    }

    /**
     * Obtém o controlador de cooldowns de um jogador.
     *
     * @param player O jogador para o qual o controlador será retornado.
     * @return O controlador de cooldowns do jogador.
     */
    public static CooldownController getCooldownController(Player player) {
        return CONTROLLER.computeIfAbsent(player.getUniqueId(), uuid -> new CooldownController());
    }
}

File: src/main/java/com/github/zyypj/tadeuBooter/api/minecraft/cooldown/CooldownController.kt
package com.github.zyypj.tadeuBooter.api.minecraft.cooldown

import org.bukkit.entity.Player
import java.util.HashMap
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * Controlador de Cooldowns para jogadores no Bukkit.
 * Permite criar, verificar e gerenciar cooldowns associados a jogadores.
 */
class CooldownController private constructor() {

    private val cooldowns: HashMap<String, Long> = HashMap()

    /**
     * Cria um cooldown associado a uma chave.
     * @param key A chave do cooldown.
     * @param seconds A duração do cooldown em segundos.
     */
    fun createCooldown(key: String, seconds: Int) {
        val expiresAt = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(seconds.toLong())
        cooldowns[key] = expiresAt
    }

    /**
     * Remove um cooldown associado a uma chave.
     * @param key A chave do cooldown a ser removido.
     */
    fun deleteCooldown(key: String) {
        cooldowns.remove(key)
    }

    /**
     * Obtém o tempo restante de um cooldown em milissegundos ou 0 se não existir.
     * @param key A chave do cooldown.
     * @return Tempo restante em milissegundos.
     */
    fun getCooldown(key: String): Long {
        val expiresAt = cooldowns[key] ?: return 0
        return expiresAt - System.currentTimeMillis()
    }

    /**
     * Verifica se um cooldown ainda está ativo.
     * @param key A chave do cooldown.
     * @return true se ativo, false caso contrário.
     */
    fun isInCooldown(key: String): Boolean = getCooldown(key) > 0

    companion object {
        private val controllers: HashMap<UUID, CooldownController> = HashMap()

        /**
         * Obtém ou cria o controlador de cooldowns de um jogador.
         * Em Java, pode chamar diretamente como CooldownController.getCooldownController(player).
         */
        @JvmStatic
        fun getCooldownController(player: Player): CooldownController =
            controllers.getOrPut(player.uniqueId) { CooldownController() }

        /**
         * Remove o controlador de cooldowns associado a um jogador.
         * Em Java, pode chamar diretamente como CooldownController.removeCooldownController(player).
         */
        @JvmStatic
        fun removeCooldownController(player: Player) {
            controllers.remove(player.uniqueId)
        }
    }
}